R/kenwardroger.R

```diff
@@ -9,98 +11,118 @@
-  if (object$tmb_data$reml != 1) {
-    stop("Kenward-Roger is only for REML!")
+h_df_md_kr <- function(object, contrast, linear = FALSE) {
+  if (component(object, "reml") != 1) {
+    stop("Kenward-Roger is only for REML")
   }
   kr_comp <- object$kr_comp
-  w <- solve(object$tmb_obj$he(object$theta_est))
+  w <- component(object, "theta_vcov")
   v_adj <- object$beta_vcov_adj
-  df <- h_kr_df(object$beta_vcov, contrast, w, kr_comp$P)
-  vsolve <- solve(contrast %*% v_adj %*% t(contrast))
-  f_statistic <- 1 / nrow(contrast) * object$beta_est %*% t(contrast) %*% vsolve %*% contrast %*% object$beta_est
+  df <- h_kr_df(v0 = object$beta_vcov, l = contrast, w = w, p = kr_comp$P)
+  vsolve <- solve(h_quad_form_mat(contrast, v_adj))
+  beta_contrast <- component(object, "beta_est") %*% t(contrast)
+  f_statistic <- 1 / nrow(contrast) * h_quad_form_mat(beta_contrast, vsolve)
   f_star <- f_statistic * df$lambda
-  ret <- list(
+  list(
     num_df = nrow(contrast),
     num_df = nrow(contrast),
     denom_df = df$m,
     f_stat = f_star[1, 1],
     p_val = pf(f_star[1, 1], nrow(contrast), df$m, lower.tail = FALSE)
   )
-  ret$sqrt <- sqrt(contrast %*% v_adj %*% t(contrast))
-  ret
 }
 
 #' Calculation of Kenward-Roger Degrees of Freedom for One-Dimensional Contrast
+#'
 #' @param object `mmrm` object
 #' @param contrast contrast vector
 #' @param linear `logical`\cr whether to use linear Kenward-Roger approximation
+#'
 #' @return List with `est`, `se`, `df`, `t_stat` and `p_val`.
+#'
 #' @keywords internal
-df_1d_kr <- function(object, contrast, linear = FALSE) {
-  if (object$tmb_data$reml != 1) {
+h_df_1d_kr <- function(object, contrast, linear = FALSE) {
+  if (component(object, "reml") != 1) {
     stop("Kenward-Roger is only for REML!")
   }
   assert_numeric(contrast, len = length(component(object, "beta_est")))
   est <- sum(contrast * component(object, "beta_est"))
-  kr_comp <- object$kr_comp
-  w <- solve(object$tmb_obj$he(object$theta_est))
+  w <- component(object, "theta_vcov")
   v_adj <- object$beta_vcov_adj
-  df <- h_kr_df(object$beta_vcov, matrix(contrast, nrow = 1), w, kr_comp$P)
-  se <- sqrt(contrast %*% v_adj %*% contrast)[1, 1]
+  df <- h_kr_df(v0 = object$beta_vcov, l = matrix(contrast, nrow = 1), w = w, p = object$kr_comp$P)
+  se <- sqrt(h_quad_form_vec(contrast, v_adj))
+
+  t_stat <- abs(est) / se
+  p_val <- pt(t_stat, df$m, lower.tail = FALSE) * 2
 
-  ret <- list(
+  list(
     est = est,
     se = se,
-    df = df$m
+    df = df$m,
+    t_stat = t_stat,
+    p_val = p_val
   )
-  t_stat <- abs(est) / se
-  ret$t_stat <- t_stat
-  ret$p_val <- pt(t_stat, df$m, lower.tail = FALSE) / 2
-  ret
 }
 
-#' obtain the adjusted Kenward-Roger degree of freedom
+#' Obtain the Adjusted Kenward-Roger Degree of Freedom
+#'
 #' @param v0 unadjusted covariance matrix
 #' @param l linear combination matrix
 #' @param w hessian matrix
 #' @param p P matrix from `h_get_kr_comp`
+#'
 #' @return List with `m` and `lambda`
```

Comment:
```suggestion
#' @return List with `m` and `lambda`.
```
please also explain what is `m` and `lambda` e.g. in parentheses after them